api/src/wfl/module/covid.clj

```diff
@@ -93,6 +93,98 @@
   "Use `tx` to load the workload sink with `sink_type`."
   (fn [tx workload] (:sink_type workload)))
 
+;; validations
+(defmulti throw-when-malformed-source-request!
+  "Validate the source of a `request`"
+  (fn [source] (:name source)))
+
+(defmulti throw-when-malformed-executor-request!
+  "Validate the executor of a `request`"
+  (fn [executor] (:name executor)))
+
+(defmulti throw-when-malformed-sink-request!
+  "Validate the sink of a `request`"
+  (fn [sink] (:name sink)))
+
+;; Workload Functions
+(def ^:private dataset-table-name "flowcells")
+(def ^:private dataset-column-name "run_date")
+(def ^:private source-name "Terra DataRepo")
+(def ^:private executor-name "Terra")
+(def ^:private sink-name "Terra Workspace")
+
+(defn ^:private add-workload-record
+  "Use `tx` to create a workload `record` for `request` and return the id of the
+   new workload."
+  [tx request]
+  (letfn [(combine-labels [labels]
+            (->> (mapv request [:pipeline :project])
+                 (map str ["pipeline:" "project:"])
+                 (concat labels)
+                 set
+                 sort
+                 vec))]
+    (-> (update request :labels combine-labels)
+        (select-keys [:creator :watchers :labels :project])
+        (merge (select-keys (wfl/get-the-version) [:commit :version]))
+        (assoc :executor "" :output "" :release "" :wdl "" :uuid (UUID/randomUUID))
+        (->> (jdbc/insert! tx :workload) first :id))))
+
+(defn throw-unless-column-exists
+  "Throw or return the column from `table`"
+  [dataset table]
+  (let [[result & more :as all] (-> table
+                                    (get-in [:columns])
+                                    (->> (filter (comp #{dataset-column-name} :name))))]
```

Comment:
This column name is the `{:keys [column] :as source-request}`

Reply:
updated